numpy/lib/nanfunctions.py

```diff
@@ -640,8 +641,15 @@ def nansum(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
     nan
 
     """
-    a, mask = _replace_nan(a, 0)
-    return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims)
+    try:
+        a = np.asanyarray(a)
+        return np.sum(a, axis=axis, dtype=dtype, out=out, keepdims=keepdims,
+                      where=_nan_mask(a))
+    except TypeError:
+        # retain old mask handling for np.matrix,
```

Comment:
It seems strange to add work-arounds for code that we are deprecating. I think at least a deprecation-warning might be reasonable. I'll ping the mailing list.